pkg/scheduler/cron.go

```diff
@@ -14,45 +14,168 @@
 package scheduler
 
 import (
+	"context"
+	"sync"
 	"time"
 
+	"github.com/go-logr/zapr"
+	"github.com/pingcap/log"
+	perr "github.com/pkg/errors"
 	cron "github.com/robfig/cron/v3"
+	"go.uber.org/zap"
 
 	"github.com/chaos-mesh/chaosd/pkg/core"
 )
 
 type Scheduler struct {
 	*cron.Cron
+	expStore    core.ExperimentStore
+	expRunStore core.ExperimentRunStore
+	cronStore   CronStore
 }
-
 type CronJob struct {
-	experiment core.Experiment
-	run        func()
+	// immutable fields
+	scheduler   *Scheduler
+	experiment  *core.Experiment
+	attackFunc  func() error
+	recoverFunc func() error
+
+	// mutable fields protected by sync.Locker
+	waitForRecovery bool
+	lock            sync.Mutex
+}
+
+func hasCronDurationExceeded(startedAt time.Time, duration time.Duration) bool {
+	return time.Until(startedAt.Add(duration)).Milliseconds() < 0
+}
+
+func (cj *CronJob) RecoverRun(expRun *core.ExperimentRun) {
+	defer cj.setWaitForRecovery(false)
+	log.Info("recovering attack on exp run", zap.String("expRunUID", expRun.UID))
+	if err := cj.recoverFunc(); err != nil {
+		log.Warn("recovery failed", zap.Error(err))
+	} else {
+		if err := cj.scheduler.expRunStore.Update(context.Background(), expRun.UID, core.RunRecovered, ""); err != nil {
+			log.Error("failed to update in DB", zap.Error(err))
+		}
+	}
 }
 
-// TODO: on running, create new experiment run
-// TODO: write tests for it
-func (cj CronJob) Run() {
-	cj.run()
+func (cj *CronJob) assertNoRecovery() bool {
+	cj.lock.Lock()
+	waitingForRecovery := cj.waitForRecovery
+	cj.lock.Unlock()
+	return !waitingForRecovery
+}
+
+func (cj *CronJob) setWaitForRecovery(waitForRecovery bool) {
+	cj.lock.Lock()
+	cj.waitForRecovery = waitForRecovery
+	cj.lock.Unlock()
+}
+
+// Run implements cron.Job interface, used when scheduling cron jobs
+func (cj *CronJob) Run() {
+	if !cj.assertNoRecovery() {
+		log.Info("skipping scheduled execution of attack since recovery in progress", zap.String("expId", cj.experiment.Uid))
+		return
+	}
+
+	var newRun *core.ExperimentRun
+	var recoverTimer *time.Timer
+	defer func() {
+		var updErr error
+		if panicErr, ok := recover().(error); panicErr != nil && ok {
```

Comment:
I run a scheduler job on stress attack, and the log print `scheduled run success`, but I didn't find the process of stress-ng, at last I find it is because some function in macOS is not implemented.

seems that here should change to `if panicErr := recover(); panicErr != nil {`

Reply:
@WangXiangUSTC not sure what u mean, the error is shown in the logs: 
```
[2021/05/17 12:09:00.010 +05:30] [ERROR] [cron.go:89] ["scheduled run errored"] [expId=1c421bdc-5841-4849-ac78-86a6637fcc7c] [error="attack failed: exec: \"stress-ng\": executable file not found in $PATH"] [errorVerbose="exec: \"stress-ng\": executable file not found in $PATH\nattack failed"] [stack="github.com/pingcap/log.Error\n\t/home/shivanshs9/go/pkg/mod/github.com/pingcap/log@v0.0.0-20200117041106-d28c14d3b1cd/global.go:42\ngithub.com/chaos-mesh/chaosd/pkg/scheduler.(*CronJob).Run.func1\n\t/home/shivanshs9/projects/chaos/chaosd/pkg/scheduler/cron.go:89\nruntime.gopanic\n\t/usr/lib/go/src/runtime/panic.go:965\ngithub.com/chaos-mesh/chaosd/pkg/scheduler.(*CronJob).Run\n\t/home/shivanshs9/projects/chaos/chaosd/pkg/scheduler/cron.go:137\ngithub.com/robfig/cron/v3.SkipIfStillRunning.func1.1\n\t/home/shivanshs9/go/pkg/mod/github.com/robfig/cron/v3@v3.0.1/chain.go:85\ngithub.com/robfig/cron/v3.FuncJob.Run\n\t/home/shivanshs9/go/pkg/mod/github.com/robfig/cron/v3@v3.0.1/cron.go:136\ngithub.com/robfig/cron/v3.(*Cron).startJob.func1\n\t/home/shivanshs9/go/pkg/mod/github.com/robfig/cron/v3@v3.0.1/cron.go:312"]
```

and also updated in DB:
```
[{"id":296,"uid":"2fcb6147-10a8-4f7f-8675-7af19192a687","start_at":"2021-05-17T12:09:00.0022572+05:30","finished_at":"0001-01-01T00:00:00Z","status":"failed","error":"attack failed: exec: \"stress-ng\": executable file not found in $PATH","ExperimentID":62,"experiment":{"id":62,"uid":"1c421bdc-5841-4849-ac78-86a6637fcc7c","created_at":"2021-05-17T12:08:33.237843307+05:30","updated_at":"2021-05-17T12:08:33.249153029+05:30","status":"scheduled","error":"","kind":"stress","action":"mem","recover_command":"{\"schedule\":\"*/1 * * * *\",\"duration\":\"30s\",\"action\":\"mem\",\"kind\":\"stress\",\"Load\":0,\"Workers\":2,\"Size\":\"100m\",\"Options\":null,\"StressngPid\":0}","launch_mode":"svr"}},{"id":297,"uid":"fc00dc03-c299-4e0f-8c31-e07608fa8874","start_at":"2021-05-17T12:10:00.001614857+05:30","finished_at":"0001-01-01T00:00:00Z","status":"failed","error":"attack failed: exec: \"stress-ng\": executable file not found in $PATH","ExperimentID":62,"experiment":{"id":62,"uid":"1c421bdc-5841-4849-ac78-86a6637fcc7c","created_at":"2021-05-17T12:08:33.237843307+05:30","updated_at":"2021-05-17T12:08:33.249153029+05:30","status":"scheduled","error":"","kind":"stress","action":"mem","recover_command":"{\"schedule\":\"*/1 * * * *\",\"duration\":\"30s\",\"action\":\"mem\",\"kind\":\"stress\",\"Load\":0,\"Workers\":2,\"Size\":\"100m\",\"Options\":null,\"StressngPid\":0}","launch_mode":"svr"}}]
```

Reply:
I mean it can't handle the panic error, you can test it by updating one `Attack` function to panic, and check it

Reply:
@WangXiangUSTC ah got it, you're referring to this: https://github.com/chaos-mesh/chaos-mesh/blob/master/pkg/bpm/build_darwin.go

Technically the issue is that the panic doesn't throw an `error` instance, but I understand changing it would block this PR. No problem, I would handle both error and string in recover

Reply:
@WangXiangUSTC fixed now